integration-test-groups/foundation/variables/src/main/java/org/apache/camel/quarkus/variables/it/VariablesResource.java

```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Path("/removeGlobalVariable")
+    @POST
+    public String removeGlobalVariable(String body) throws Exception {
+        MockEndpoint mid = context.getEndpoint("mock:removeGlobalVariableMid", MockEndpoint.class);
+        mid.expectedMessageCount(1);
+
+        producerTemplate.requestBody("direct:removeGlobalVariableStart", body, String.class);
+
+        // make sure we got the message
+        mid.assertIsSatisfied();
+
+        // lets get the variable value
+        List<Exchange> exchanges = mid.getExchanges();
+        Exchange midExchange = exchanges.get(0);
+
+        return midExchange.getVariable(VariablesRoutes.VARIABLE_NAME, String.class) + ","
+                + context.getVariable(VariablesRoutes.VARIABLE_NAME, String.class);
+    }
+
+    @Path("/customGlobalRepository")
+    @POST
+    public Response customGlobalRepository(String body) throws Exception {
+        //        context.getRegistry().bind(GLOBAL_VARIABLE_REPOSITORY_ID, new MyGlobalRepo());
+
+        context.getRouteController().startRoute("customGlobalRepository");
+
+        MockEndpoint end = context.getEndpoint("mock:setGlobalCustomEnd", MockEndpoint.class);
+        end.expectedMessageCount(1);
+
+        producerTemplate.requestBody("direct:setGlobalCustomStart", body, String.class);
+
+        // make sure we got the message
+        end.assertIsSatisfied();
+
+        // lets get the variable value
+        List<Exchange> exchanges = end.getExchanges();
+        Exchange exchange = exchanges.get(0);
+
+        String resp = exchange.getVariable(VariablesRoutes.VARIABLE_NAME, String.class) + ","
+                + context.getVariable(VariablesRoutes.VARIABLE_NAME, String.class);
+        return Response.ok().entity(resp).build();
+    }
+
+    @Path("/convert")
+    @POST
+    public String convert(String body) throws Exception {
+        MockEndpoint end = context.getEndpoint("mock:convertEnd", MockEndpoint.class);
+        end.expectedMessageCount(1);
+
+        producerTemplate.requestBody("direct:convertStart", body, String.class);
+        //        fluentProducerTemplate.to("direct:convertStart").withVariable(VariablesRoutes.VARIABLE_NAME, body).send();
+
+        // make sure we got the message
+        end.assertIsSatisfied();
+
+        List<Exchange> exchanges = end.getExchanges();
+        Exchange exchange = exchanges.get(0);
+
+        return exchange.getVariable(VariablesRoutes.VARIABLE_NAME, String.class);
+    }
+
+    @Path("/filter/{city}")
+    @POST
+    public Response filter(String body, @PathParam("city") String city) throws Exception {
+        MockEndpoint end = context.getEndpoint("mock:filterEnd", MockEndpoint.class);
+        end.expectedMessageCount(1);
+
+        producerTemplate.requestBodyAndHeader("direct:filterStart", body, "city", city, String.class);
+
+        try {
+            Exchange exhange = end.assertExchangeReceived(0);
+            end.reset();
```

Comment:
Didn't we have some kind of race condition issue generating flaky tests in the past ? A bit like 2 threads invoking the filter method concurrently and then 1 thread call reset while the other thread had the exchange.

Not sure it applies here anyway, still it would it be good to find another way to test without reset ?

Reply:
If there is a possibility of flakiness, I think that I refactor the test class and ensure that calls to this resource  are synchronized (by joining 2 test methods into 1)

Reply:
Also, I remember a technic where we had one distinct mock per expected message. So that you never call reset... That was far, hard to remember if this applies to the case at hand :)

Reply:
I skipped `reset()`  by rearranging test method differently.